docs/source/checkpoints/dist_ckpt.rst

```diff
@@ -0,0 +1,392 @@
+The key in the state dict is ephemeral, while the `ShardedTensor.key` is used to identify the tensor in the checkpoint.
+
+Example:
+
+.. code-block:: python
+    import torch
+
+    from megatron.core import dist_checkpointing
+
+    # Checkpoint saved with some key in the state dict that is eventually ignored
+    model = ...
+    ckpt_dir = ...
+    sharded_state_dict = {
+        'ignored': dist_checkpointing.ShardedTensor('tensor-A', ...)
+    }
+    dist_checkpointing.save(sharded_state_dict, ckpt_dir)
+
+    # During loading, all that matters is the ShardedTensor.key.
+    sharded_state_dict = {
+        'different-key': dist_checkpointing.ShardedTensor('tensor-A', ...)
+    }
+    loaded_state_dict = dist_checkpointing.load(sharded_state_dict, ckpt_dir)
+    assert 'ignored' not in loaded_state_dict
+    assert 'tensor-A' not in loaded_state_dict
+    assert isinstance(loaded_state_dict['different-key'], torch.Tensor)
+
+    # The key in the state dict is important only from the subsequent `model.load_state_dict`
+    # that usually happens after `dist_checkpointing.load` - the state dict must have
+    # the structure and keys corresponding to the model structure and submodule names.
+    model.load_state_dict(loaded_state_dict)
+
+ShardedTensor
+-------------
+It's the primary use case of distributed checkpointing - tensors sharding.
+Allows to define how PyTorch tensors are sharded across the workload.
+See `Tensors transformations`_ section for more details on ShardedTensors.
+
+ShardedObject
+-------------
+Sometimes there is a need to save arbitrary objects across the ranks.
+ShardedObject allows to structure those objects into arrays of objects with a fixed `global_shape` and save/load parts of the arrays on specific ranks.
+
+ShardedTensorFactory
+--------------------
+This class allows to defer tensors transformations until the actual saving.
+A factory can expand a tensor into an arbitrary sub state dict (including all supported entities listed above).
+The need for such deferral will be explained in the `Tensors transformations`_ section.
+
+LocalNonpersitentObject
+-----------------------
+This is a simple wrapper that allows to express the fact that the object wrapped with this class should end up in the final loaded state dict during loading.
+During saving such objects are ignored.
+
+Arbitrary object
+----------------
+All objects different than dicts, lists and the instances of the classes listed above are treated as "common" objects.
+
+During saving, all such objects in the sharded state dict passed to `dist_checkpointing.save` are assumed to be duplicated across ranks and therefore saved only by a single coordinator rank (rank 0).
+
+During loading, all such objects in the sharded state dict passed to `dist_checkpointing.load` are simply ignored - the loaded state dict contains only "common" objects that are were actually saved in the checkpoint.
```

Comment:
Objects that are not dictionaries, lists, or instances of the previously mentioned classes are categorized as arbitrary objects or “common” objects. When these objects are included in the sharded state dictionary and passed to the dist_checkpointing.save function during the saving process, it is assumed that they are replicated across all ranks. Consequently, they are saved solely by the coordinator rank, specifically rank 0.